llms.md

```diff
@@ -0,0 +1,1039 @@
+*   `#[reducer(client_connected)]`: Runs when any distinct client connection (e.g., WebSocket, HTTP call) is established. Failure disconnects the client. `ctx.connection_id` is guaranteed to be `Some(...)` within this reducer.
+*   `#[reducer(client_disconnected)]`: Runs when any distinct client connection terminates. Failure is logged but does not prevent disconnection. `ctx.connection_id` is guaranteed to be `Some(...)` within this reducer.
+
+These reducers cannot take arguments beyond `&ReducerContext`.
+
+```rust
+// Example init reducer was shown previously
+#[reducer(client_connected)]
+pub fn handle_connect(ctx: &ReducerContext) {
+    log::info!("Client connected: {}", ctx.sender);
+    // ... setup initial state for ctx.sender ...
+}
+
+#[reducer(client_disconnected)]
+pub fn handle_disconnect(ctx: &ReducerContext) {
+    log::info!("Client disconnected: {}", ctx.sender);
+    // ... cleanup state for ctx.sender ...
+}
+```
+
+#### Filtering and Deleting with Indexes
+
+SpacetimeDB provides powerful ways to filter and delete table rows using B-tree indexes. The generated accessor methods accept various argument types:
+
+*   **Single Value:** Pass a reference (`&T`) for the indexed column type.
+*   **Ranges:** Use Rust's range syntax (`start..end`, `start..=end`, `..end`, `..=end`, `start..`). Values can be owned or references.
+*   **Multi-Column Indexes:** Pass a tuple containing values or ranges for each indexed column. The types must match the column order in the index definition. You can filter on a prefix of the index columns.
+
+```rust
+#[table(name = "points", index(name = "idx_xy", btree(columns = [x, y])))]
+pub struct Point { x: i64, y: i64 }
+#[table(name = "items", index(btree(columns = [name])))]
+pub struct Item { #[primary_key] id: u32, name: String }
+
+#[reducer]
+fn index_operations(ctx: &ReducerContext) {
+    // Example: Find items named "Sword"
+    for item in ctx.db.items().name().filter("Sword") {
+        // ...
+    }
+
+    // Example: Delete points where x is between 5 (inclusive) and 10 (exclusive)
+    let num_deleted = ctx.db.points().idx_xy().delete(5i64..10i64);
+    log::info!("Deleted {} points", num_deleted);
+
+    // Example: Find points where x = 3 and y >= 0
+    for point in ctx.db.points().idx_xy().filter((3i64, 0i64..)) {
+        // ...
+    }
+
+    // Example: Delete all points where x = 7 (filtering on index prefix)
+    ctx.db.points().idx_xy().delete(7i64);
+
+    // Using references
+    let min_x = 100i64;
+    let max_x = 200i64;
+    for point in ctx.db.points().idx_xy().filter(&min_x..=&max_x) {
+         // ...
+    }
+}
+```
+
+##### Using `try_insert()`
+
+Instead of `insert()`, which panics or throws if a constraint (like a primary key or unique index violation) occurs, Rust modules can use `try_insert()`. This method returns a `Result<RowType, String>`, allowing you to gracefully handle potential insertion failures.
```

Comment:
The error variant returned from `try_insert` is not `String`, it's `TryInsertError<Tbl>`, where `Tbl` is the table handle type. `TryInsertError` has variants `UniqueConstraintViolation(Tbl::UniqueConstraintViolation)` and `AutoIncOverflow(Tbl::AutoIncOverflow)`. Those associated types may be uninhabited if the table does not have the associated constraint.